mars/services/task/supervisor/task_manager.py

```diff
@@ -0,0 +1,464 @@
+# Copyright 1999-2020 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import sys
+from collections import defaultdict, deque
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Iterable, Optional
+
+from .... import oscar as mo
+from ....core import TileableGraph, ChunkGraph, ChunkGraphBuilder, get_tiled
+from ....utils import build_fetch
+from ...cluster.api import ClusterAPI
+from ..analyzer import GraphAnalyzer
+from ..core import Task, TaskResult, TaskStatus, Subtask, SubtaskResult, \
+    SubTaskStatus, SubtaskGraph, new_task_id
+from ..errors import TaskNotExist
+
+
+class TaskProcessor:
+    __slots__  = '_task', 'tileable_graph'
+
+    def __init__(self,
+                 task: Task):
+        self._task = task
+        self.tileable_graph = task.tileable_graph
+
+    def optimize(self) -> TileableGraph:
+        """
+        Optimize tileable graph.
+
+        Returns
+        -------
+        optimized_graph: TileableGraph
+
+        """
+        # TODO(qinxuye): enable optimization
+        return self.tileable_graph
+
+    def tile(self, tileable_graph: TileableGraph) -> Iterable[ChunkGraph]:
+        """
+        Generate chunk graphs
+
+        Returns
+        -------
+        chunk_graph_generator: Generator
+             Chunk graphs.
+        """
+        # TODO(qinxuye): integrate iterative chunk graph builder
+        chunk_graph_builder = ChunkGraphBuilder(
+            tileable_graph, fuse_enabled=self._task.fuse_enabled)
+        yield from chunk_graph_builder.build()
+
+
+class BandQueue:
+    def __init__(self):
+        self._queue = deque()
+        self._has_data = asyncio.Event()
+
+    def put(self, subtask: Optional[Subtask]):
+        self._queue.appendleft(subtask)
+        self._has_data.set()
+
+    def get(self) -> Subtask:
+        subtask = self._queue.popleft()
```

Comment:
Is the `BandQueue` a queue? I found
- `popleft()` in `get()`
- `appendleft()` in `put()`

Reply:
It's actually a FILO queue.

Reply:
At first,  I used `asyncio.LifoQueue`, but later, I found that what I need is to await the queue. `BandQueue` is not the ultimate implementation, it's gonna be moved into `scheduling` service, and in it, it will become a heapq finally, now, it's only an intermediate implementation before ready of scheduling.

Reply:
Thank you for the explanation.